cypress-io/cypress
# feat: Add support for optional env key to browser launch options

### User facing changelog

n/a

### Additional details

Related PR: https://github.com/cypress-io/cypress/pull/22250

In building out support to [integrate replay with cypress](https://github.com/replayio/replay-cli/tree/main/packages/cypress),  I discovered there isn't a way to set environment variables for the browser without putting the onus on the end user. Further, since the browser launch seems to happen in a different process than the plugin, I couldn't inject those values into `process.env` and have them picked up by the existing logic.

This adds support for an optional `env` key on the launch options passed to `before:browser:launch` which is then merged with the `defaultBrowserEnv` and `process.env`. I placed it between those two so the default browser family config (`defaultBrowserEnv`) would be overwritten by browser instance config (`browser.env`) which would be overwritten by the current `process.env`.

### Steps to test


* Create a new cypress project
* Add a `before:browser:launch` handler in `setupNodeEvents` to add an env variable
* Verify that the browser receives that env variable. I added a debug entry to log the environment which is probably sufficient for testing this change.

### How has the user experience changed?


### PR Tasks


- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/4730
